drafts/npm.md

```diff
@@ -0,0 +1,472 @@
+  sabotage](https://blog.sonatype.com/npm-libraries-colors-and-faker-sabotaged-in-protest-by-their-maintainer-what-to-do-now):
+  without a reproducible installation, developers installed and ran a newly
+  published (compromised) version of the dependency on their CI/CD system,
+  giving an attacker immediate code execution on their infrastructure.
+
+- Detecting package corruptions before installation, for example due to RAM
+  corruption.
+
+- Mitigating package mutability introduced by mis-configured registries that
+  proxy packages from public registries. Note: Versions are immutable in
+  principle, but immutability is enforced by the registry.
+
+There are two ways to reliably achieve a reproducible installation: vendoring
+dependencies and pinning by hash.
+
+#### Vendoring dependencies
+
+Vendoring dependencies means keeping a local copy of all the dependencies
+(direct and transitive) in the repository. Note that updates to dependencies
+may generate large pull requests in the repository's commit history.
+
+#### Hash pinning
+
+Hash pinning informs the package manager the expected hash for each dependency,
+without trusting the registries. The package manager then verifies, during each
+installation, that the hash of each dependency remains the same.
+
+Npm provides two options to achieve hash pinning.
+
+##### Package-lock.json
+
+The
+[package-lock.json](https://docs.npmjs.com/cli/v6/configuring-npm/package-lock-json)
+contains ***all*** the dependencies (direct and indirect) along with a
+cryptographic hash of their content:
+
+```json
+{
+  "name": "A",
+  "version": "0.1.0",
+  ...metadata fields...
+  "dependencies": {
+    "B": {
+      "version": "0.0.1",
+      "resolved": "https://registry.npmjs.org/B/-/B-0.0.1.tgz",
+      "integrity": "sha512-DeAdb33F+"
+      "dependencies": {
+        "C": {
+          "version": "git://github.com/org/C.git#5c380ae319fc4efe9e7f2d9c78b0faa588fd99b4"
+        }
+      }
+    }
+  }
+}
+```
+
+The package-lock.json file is a ***snapshot of an installation*** that allows
+later reproduction of the same installation. As such, the lock file is
+generated or updated via the various commands that install packages, e.g., `npm
+install`. If some dependencies are missing or not pinned by hash (e.g.,
+`integrity` field is not present), the installation will patch the lock file to
+reflect the installation.
+
+The lock file ***cannot*** be uploaded to a registry, which means that
+consumers who locally install the package via `npm install` may [see different
+dependency
+versions](https://dev.to/saurabhdaware/but-what-the-hell-is-package-lock-json-b04)
+than the repo owners used during testing. Using package-lock.json is akin to
+dynamic linking for low-level programming languages: the loader will resolve
+dependencies at load time using libraries available on the system. Using this
+lock file leaves the task of deciding dependencies' versions to use to the
+package consumer.
```

Comment:
this fact is why i believe it's a *harmful* practice for published packages to use dev-only lockfiles: because it gives them a falsely insulated experience, while their users are exposed to breakage and vulns in transitive deps.